feat(containers): implement global network tethering for playwright server

aslushnikov opened the pull request:
This patch implements a new mode of network tethering for Playwright server & its clients.
With this patch:
- playwright server could be launched with the `--browser-proxy-mode=tether` flag to engage in the new mode
- a new type of client, "Network Tethering Client" can connect to the server to provide network traffic to the browsers
- all clients that connect to the server with the `x-playwright-proxy: *` header will get traffic from the "Network Tethering Client"

This patch also adds an environment variable `PW_OWNED_BY_TETHER_CLIENT`. With this env, playwright server will auto-close when the network tethering client disconnects. It will also auto-close if the network client does not connect to the server in the first 10 seconds of the server existence. This way we can ensure that `npx playwright docker start` blocks terminal & controls the lifetime of the started container.

dgozman commented on packages/playwright-test/src/plugins/dockerPlugin.ts:
```diff
@@ -33,20 +33,13 @@ export const dockerPlugin: TestRunnerPlugin = {
     }
     println(colors.dim(`View screen: ${info.vncSession}`));
     println('');
     process.env.PW_TEST_CONNECT_WS_ENDPOINT = info.wsEndpoint;
```
- this will be `PLAYWRIGHT_ENDPOINT`, vscode will set it to it's own endpoint and docker plugin will not engage;
- vscode will set `reuse` header;
- `run-server --controlled-by-tether-client` to make server use global socks proxy for all clients;
- `ENV=SERVER_AUTO_DIE run-server` to auto-die after tether disconnects.

aslushnikov commented on packages/playwright-core/src/remote/playwrightServer.ts:
```diff
@@ -132,16 +145,23 @@ export class PlaywrightServer {
       else
         semaphore.setMax(this._options.maxConcurrentConnections);
 
+      const autoCloseIfConnectionDies = request.headers['x-playwright-auto-close-on-disconnect'];
```
shaped this in a CLI flag

aslushnikov commented on packages/playwright-core/src/remote/playwrightServer.ts:
```diff
@@ -107,7 +114,13 @@ export class PlaywrightServer {
+        socksProxy = 'create-new';
+      else if (this._options.enableSocksProxy && proxyValue === 'use-global-network-tethering')
+        socksProxy = 'use-global';
+      else if (this._options.enableSocksProxy && proxyValue === 'bind-global-network-tethering')
```
We decided to have a server mode that will forward all browser proxies to the same socks proxy provided by a tether client.

aslushnikov commented on packages/playwright-core/src/remote/playwrightConnection.ts:
```diff
@@ -139,6 +144,10 @@ export class PlaywrightConnection {
+      playwright.options.socksProxyPort = this._globalSocksProxy?.port();
+    else
+      playwright.options.socksProxyPort = undefined;
     this._cleanups.push(() => gracefullyCloseAll());
```
this went away

dgozman commented on packages/playwright-core/src/server/transport.ts:
```diff
@@ -88,13 +88,13 @@ export class WebSocketTransport implements ConnectionTransport {
       maxPayload: 256 * 1024 * 1024, // 256Mb,
       // Prevent internal http client error when passing negative timeout.
-      handshakeTimeout: Math.max(progress.timeUntilDeadline(), 1),
+      handshakeTimeout: Math.max(progress?.timeUntilDeadline() ?? 10_000, 1),
```
```suggestion
      handshakeTimeout: Math.max(progress?.timeUntilDeadline() ?? 30_000, 1),
```

mxschmitt commented on packages/playwright-core/src/containers/docker.ts:
```diff
@@ -228,14 +235,26 @@ export async function ensurePlaywrightContainerOrDie(): Promise<ContainerInfo> {
     }
   }
 
+  const env: any = {
```
```suggestion
  const env: Record<string, any> = {
```

aslushnikov committed:
feat(containers): implement global network tethering

This patch implements a new mode of network tethering for Playwright
server & its clients.

With this patch:
- playwright server could be launched with the
  `--browser-proxy-mode=tether` flag to engage in the new mode
- a new type of client, "Network Tethering Client" can connect to the
  server to provide network traffic to the browsers
- all clients that connect to the server with the `x-playwright-proxy:
  *` header will get traffic from the "Network Tethering Client"

This patch also adds an environment variable
`PW_OWNED_BY_TETHER_CLIENT`. With this env, playwright server will
auto-close when the network tethering client disconnects. It will also
auto-close if the network client does not connect to the server in the
first 10 seconds of the server existence. This way we can ensure that
`npx playwright docker start` blocks terminal & controls the lifetime of
the started container.

aslushnikov committed:
fix tests

aslushnikov committed:
rename

aslushnikov committed:
fix type
